pwndbg/heap/ptmalloc.py

```diff
@@ -1608,218 +1597,189 @@ def thread_cache(self):
             self._thread_cache = self.tcache_perthread_struct(thread_cache_via_config)
             return self._thread_cache
         elif thread_cache_via_symbol:
-            if pwndbg.gdblib.symbol.static_linkage_symbol_address("tcache"):
-                # If the symbol is static-linkage symbol, we trust it.
-                thread_cache_struct_addr = pwndbg.gdblib.memory.u(thread_cache_via_symbol)
-                self._thread_cache = self.tcache_perthread_struct(thread_cache_struct_addr)
-                return self._thread_cache
             # Check &tcache is nearby TLS base or not to avoid false positive.
-            tls_base = pwndbg.gdblib.tls.address
-            if tls_base:
-                if pwndbg.gdblib.arch.current in ("x86-64", "i386"):
-                    is_valid_address = 0 < tls_base - thread_cache_via_symbol < 0x250
-                else:  # elif pwndbg.gdblib.arch.current in ("aarch64", "arm"):
-                    is_valid_address = 0 < thread_cache_via_symbol - tls_base < 0x250
-
-                is_valid_address = (
-                    is_valid_address
-                    and thread_cache_via_symbol in pwndbg.gdblib.vmmap.find(tls_base)
-                )
-
-                if is_valid_address:
-                    thread_cache_struct_addr = pwndbg.gdblib.memory.u(thread_cache_via_symbol)
+            if pwndbg.gdblib.tls.is_thread_local_variable(thread_cache_via_symbol):
+                thread_cache_struct_addr = pwndbg.gdblib.memory.u(thread_cache_via_symbol)
+                if pwndbg.gdblib.vmmap.find(thread_cache_struct_addr):
                     self._thread_cache = self.tcache_perthread_struct(thread_cache_struct_addr)
                     return self._thread_cache
 
-        if self.has_tcache():
-            # Each thread has a tcache struct, and the address of the tcache struct is stored in the TLS.
+        # If target is single-threaded, then the tcache struct is at the first chunk of the heap.
+        # We try to find the address by using mp_.srck_base + 0x10 first since it's more reliable than other methods.
+        if not self.multithreaded:
+            try:
+                thread_cache_struct_addr = self.mp["sbrk_base"] + 0x10
```

Comment:
Just a minor point, `thread_cache_struct_addr` will be inaccurate if the `GLIBC_TUNABLES` environment variable is set.
Here's an example of how pwndbg can deal with this: https://github.com/pwndbg/pwndbg/blob/52a479211cf58d1bdc5b42fd50471f565e6a063c/pwndbg/heap/ptmalloc.py#L1231-L1237

Reply:
Thanks pointing it out, I'll use `get_sbrk_heap_region()` instead.
Btw [here](https://github.com/pwndbg/pwndbg/blob/d4562ea462b8013253c219ed5ddd292386be5c6c/pwndbg/heap/ptmalloc.py#L1139) also used `sbrk_base + 0x10`, should I replace it as well?

Reply:
Well spotted, yes that also seems like a good candidate for replacement. It might need some extra modifications since a `Page` object is returned.

Reply:
If there are any more improvements to be made here, please lets do it in another PR